src/main/scala/calc/Compiler.scala

```diff
@@ -61,21 +62,107 @@ object Compiler {
    *  that evaluates to the result of the tree.
    *
    *  This is the main method you have to implement.
    */
-  def compileExpr(tree: Tree): irt.Tree = {
-    // TODO
+  def compileExpr(tree: Tree, typeEnv: Map[String, irtpe.Type]): irt.Tree = {
     implicit val pos = tree.pos
+    val functionTypes:Map[String, Int] = Map("sin" -> 1,
+      "cos" -> 1,
+      "tan" -> 1,
+      "log" -> 1,
+      "pow" -> 2,
+      "abs" -> 1,
+      "max" -> 2,
+      "min" -> 2)
 
     tree match {
       case Literal(value) =>
         irt.DoubleLiteral(value)
 
-      case _ =>
-        throw new Exception(
+      case BinaryOp(op, lhs, rhs) =>
+        val jsop = op match {
+          case "+" => irt.BinaryOp.Double_+
+          case "-" => irt.BinaryOp.Double_-
+          case "*" => irt.BinaryOp.Double_*
+          case "/" => irt.BinaryOp.Double_/
+        }
+        irt.BinaryOp(jsop, compileExpr(lhs, typeEnv), compileExpr(rhs, typeEnv))
+
+      case Let(ident, value, body) =>
+        val jsV = compileExpr(value, typeEnv)
+        val jsIdent = irt.Ident(ident.name)
+        val jsId = irt.VarDef(jsIdent, jsV.tpe, mutable = false, jsV)
+        irt.Block(List(jsId, compileExpr(body, typeEnv + (ident.name -> jsV.tpe))))
+
+      case Ident(n) =>
+        try {
+          irt.VarRef(irt.Ident(n))(typeEnv(n))
+        }catch{
+          case e:Exception => throw new Exception("No such identitier!")
+        }
+
+      case If(cond, thenp, elsep) =>
+        val jsThenp = compileExpr(thenp, typeEnv)
+        val jsElsep = compileExpr(elsep, typeEnv)
+        val jsCond = compileExpr(cond, typeEnv)
+        val finalCond = irt.BinaryOp(irt.BinaryOp.!==, jsCond, irt.DoubleLiteral(0.0))
+        if (jsElsep.tpe != jsElsep.tpe){
+          throw new Exception(s"The then and also phase in the conditional statement must bse same type")
+        }
+        irt.If(finalCond, jsThenp, jsElsep)(jsThenp.tpe)
+
+      case Closure(idList, body) =>
+        val captureList = findCapture(tree).toList
+        val captureParam = captureList.map(s =>
+          irt.ParamDef(irt.Ident(s), typeEnv(s), mutable = false, rest = false))
+        val params = idList.map(s =>
+          irt.ParamDef(irt.Ident("$$"+s.name), irtpe.AnyType, mutable = false, rest = false))
+        val newTypeEnv = idList.foldLeft(typeEnv)((env: Map[String, irtpe.Type], s:Ident)
+                                                    => env+(s.name -> irtpe.DoubleType))
+        val unboxParams = idList.map(s =>
+          irt.VarDef(irt.Ident(s.name), irtpe.DoubleType, mutable = false,
+            irt.Unbox(irt.VarRef(irt.Ident("$$"+s.name))(irtpe.AnyType), 'D')))
+        val bl = irt.Block(unboxParams ++ List(compileExpr(body, newTypeEnv)))
+        irt.Closure(captureParam, params, bl, captureList.map(s => irt.VarRef(irt.Ident(s))(typeEnv(s))))
+
+      case Call(Ident(name), args) =>
+        try {
+          if (functionTypes.contains(name)) {
```

Comment:
In case `name` is defined *both* in `functionTypes` and in `typeEnv`, the one from `typeEnv` should take precedence, so that shadowing works: `let sin = 5 in sin` should work and yield 5.